ospsuite-r-specifics/r-code-structure.md

```diff
@@ -0,0 +1,250 @@
+    },
+    #' @field id The id of the .NET wrapped object. (read-only)
+    id = function(value) {
+      private$wrapReadOnlyProperty("Id", value)
+    }
+  )
+)
+```
+
+As you can see in the R simulation class, we provide access to simulation properties (like f.e. the simulation Output Schema) using the functionalities of the `DotNetWrapper`:
+
+[simulation.R](https://github.com/Open-Systems-Pharmacology/OSPSuite-R/blob/develop/R/simulation.R)
+```
+#' @field outputSchema outputSchema object for the simulation (read-only)
+outputSchema = function(value) {
+    private$readOnlyProperty("outputSchema", value, private$.settings$outputSchema)
+}
+```
+Please not that it is a requirement for these R wrapper classes to implement a meaningful print function. In our example:
+
+[simulation.R](https://github.com/Open-Systems-Pharmacology/OSPSuite-R/blob/develop/R/simulation.R)
+```
+#' @description
+#' Print the object to the console
+#' @param ... Rest arguments.
+print = function(...) {
+    private$printClass()
+    private$printLine("Number of individuals", self$count)
+    invisible(self)
+}
+```
+
+
+Many times the basic access to the object methods and properties is not sufficient, and we need further functionalities on the objects. For this we create a functions that work on that objects and pack them in separate utilities files. For our example with simulation, we have [utilities-simulation.R](https://github.com/Open-Systems-Pharmacology/OSPSuite-R/blob/develop/R/utilities-simulation.R). These utilities files contain R code that works on the created objects of the class, but also if necessary rClr calls to .NET functions that work on the objects. Note that rClr functions that just expose properties or methods of the objects do NOT belong here, but in the R wrapper class. 
+
+Please also note that per convention all functions that are only used internally by the package are named starting with a dot. For example `.runSingleSimulation` and not just `runSingleSimulation`:
+
+[utilities-simulation.R](https://github.com/Open-Systems-Pharmacology/OSPSuite-R/blob/develop/R/utilities-simulation.R)
+```
+.runSingleSimulation <- function(simulation, simulationRunOptions, population = NULL, agingData = NULL) {
+```
+
+
+
+The communication between R and .NET does not come without some overhead. This means that when we can avoid it we should. 
+
+### Tasks and task caching
+
+Often when working with objects we use tasks. Those tasks are objects defined and created on the .NET side that are reusable and can provide functionalities on other objects. They can be accessed through the [Api.cs](https://github.com/Open-Systems-Pharmacology/OSPSuite.Core/blob/develop/src/OSPSuite.R/Api.cs) of OSPSuite.Core as usual - on the OSPSuite side they are created through the IoC container. Let's see for example how we can use the HasDimension utility function of the unit wrapper class to check if a dimension (provided as a string) is supported.
```

Comment:
What is an `IoC` container? Please keep in mind that this documentation is also for R developers who might not be familiar with abbreviations use in other programming environments.

Reply:
added wikipedia link explaining IoC (Inversion of Control)